### ISSUE TITLE
Why not upgrade acorn.js 

### ISSUE THREAD

**Issue by KRISACHAN** [NeilFraser/JS-Interpreter#186]
...
I saw the lib, acorn.js's version just 0.4.1 ....
If I use the high version ecmascript code , for example :
```js
for (let i of [1, 2, 3]) {
console.log(i)
}
```
It will throw an error ...

The last version of acorn.js is 7.1.0 ,but Js-Interpreter only 0.x

Too terrible

**Comment by cpcallen** [comment]
The limiting factor is not Acorn (though updating it might not be a bad idea anyway), but that JS Intrpreter itself only understands ES5.1.  So being able to *parse* `for ... of` loops wouldn't help, because the interpreter still wouldn't know how to execute them.

Nevertheless, since no one seems to have opened a "please support ES6" issue yet, I'll repurpose your report accordingly.

**Comment by KRISACHAN** [comment]
Thank you for your answer 
I try to repackage with acorn@7.x just now .
I don't know if it's a good idea .
:)
Happy Weekend !

**Comment by cpcallen** [comment]
> I try to repackage with acorn@7.x just now.

Just make sure you specify the appropriate options to Acorn so that it will parse ES5 only.

**Comment by sag1v** [comment]
@cpcallen 
Supporting ES6+ features would be awesome, this is the only "deal breaker" for using the lib.
Is it something you are considering doing?

**Comment by cpcallen** [comment]
> Is it something you are considering doing?

This is a question for @NeilFraser—but I suspect the answer is no, because this will eventually be part of the roadmap of another project (which is actually funded, unlike this one), so in a few years we will have a much better solution.

In the mean time, I would suggest using an ES6->5 transpiler.  I have heard that Babel can do this.

**Comment by sag1v** [comment]
Thanks @cpcallen. Actually my use case for this library is to create a JavaScript "visualiser" (similar to devtools), not sure babel can help with this as it will change that actual code i want to visualise.

**Comment by KSean222** [comment]
Same here, I have tried using source maps to correct this, but they don't seem to come out correctly.

**Comment by NeilFraser** [comment]
Just to confirm, *I* have no interest in developing and maintaining an ES6+
version.  It would be a step backwards for my projects in that the download
times would increase.

However, I recognize that there is a genuine need for ES6+ support in other
projects and I'd fully support anyone who wishes to fork the project
accordingly.

Sent (awkwardly) from an Android device.

**Comment by sag1v** [comment]
@KSean222 What do you mean by 
> ... using source maps to correct this...

**Comment by KSean222** [comment]
https://developer.mozilla.org/en-US/docs/Tools/Debugger/How_to/Use_a_source_map
Babel can generate source maps when converting to ES5.

**Comment by sag1v** [comment]
>  https://developer.mozilla.org/en-US/docs/Tools/Debugger/How_to/Use_a_source_map
> Babel can generate source maps when converting to ES5.
> 
Yeah i know, but how do you think it to sync them? Sounds interesting

**Comment by KSean222** [comment]
I tried using https://github.com/mozilla/source-map @sag1v

**Comment by cheapsteak** [comment]
> > Is it something you are considering doing?
> 
> This is a question for @NeilFraser—but I suspect the answer is no, because this will eventually be part of the roadmap of another project (which is actually funded, unlike this one), so in a few years we will have a much better solution.
> 
> In the mean time, I would suggest using an ES6->5 transpiler. I have heard that Babel can do this.

What is the name of the funded project?

**Comment by cpcallen** [comment]
> What is the name of the funded project?

[Code City](http://github.com/google/CodeCity/).  But do note: "will *eventually* be *part of the roadmap*"—in its present form Code City is actually a less-complete JavaScript implementation than JS Interpreter is.

**Comment by bendavis78** [comment]
I'm very interested in this, as it would be useful for creating a client-side step debugger for [Theia IDE](theia-ide.org).

In theory, how much work are we talking to make this happen? Would it require a major re-write of the interpreter? Or is it just adding ES-6 specific features?

**Comment by cpcallen** [comment]
@bendavis78: It would be a considerable amount of work.  ES6 is quite a bit larger language than ES5.1, and ES2020 is even more so.  I don't think it necessarily requires a throw-it-away-and-start-from-scratch rewrite: it could certainly be done incrementally—but by the time it was done very little of the original code base will remain.
